### PR TITLE
chore(benchpress): add browserify bundling

### PR DESCRIPTION
**Pull request by uber5001**
Use browserify to bundle benchpress and its dependencies.

[Review on Reviewable](https://reviewable.io/reviews/angular/angular/2711)

### CODE REVIEW

File: gulpfile.js [hunk 1]
```diff
@@ -940,6 +944,25 @@ gulp.task('cleanup.builder', function(done) {
   });
 });
 
+gulp.task('benchpress.bundle.prefix', ['build.js.cjs'], function() {
+  return gulp.src(['./dist/js/cjs/benchpress/index.js'])
+      .pipe(insert.prepend('require("reflect-metadata");\n'))
```
Comment (tbosch) [inline gulpfile.js]: This should actually be moved into `modules/benchpress/index.js`

File: gulpfile.js [hunk 2]
```diff
@@ -940,6 +944,25 @@ gulp.task('cleanup.builder', function(done) {
...
+      .pipe(insert.append('\nglobal.__benchpressExports = module.exports;\n'))
```
Comment (tbosch) [inline gulpfile.js]: Move this also into `modules/benchpress/index.js`

File: gulpfile.js [hunk 3]
```diff
@@ -940,6 +944,25 @@ gulp.task('cleanup.builder', function(done) {
   });
 });
 
+gulp.task('benchpress.bundle.prefix', ['build.js.cjs'], function() {
```
Comment (tbosch) [inline gulpfile.js]: Remove this task

File: gulpfile.js [hunk 4]
```diff
@@ -940,6 +944,25 @@ gulp.task('cleanup.builder', function(done) {
...
+      .pipe(insert.prepend('require("reflect-metadata");\n'))
+      .pipe(insert.append('\nglobal.__benchpressExports = module.exports;\n'))
+      .pipe(gulp.dest('./dist/js/cjs/benchpress/'))
+})
+
+gulp.task('benchpress.bundle', ['build.js.cjs', 'benchpress.bundle.prefix'], function() {
+  var b = browserify({
+    entries: './dist/js/cjs/benchpress/index.js'
+  });
+  b.exclude('selenium-webdriver');
```
Comment (tbosch) [inline gulpfile.js]: Make this exclude generic:

Exclude all dependencies that are listed in `dist/js/cjs/benchpress/package.json` and `dist/js/cjs/angular2/package.json`, except for `reflect-metadata` and `angular2`

Please also adjust `modules/benchpress/package.json` to remove `rtts_assert`.

File: gulpfile.js [hunk 5]
```diff
@@ -940,6 +944,25 @@ gulp.task('cleanup.builder', function(done) {
...
+  return b.bundle()
```
Comment (tbosch) [inline gulpfile.js]: Create a new folder `dist/js/cjs/benchpress_bundle` with a package.json that only dependencies that are excluded and your current `bundle.js` as `index.js`. Maybe also copy over `README.md` and `docs` folder.

**Comment by tbosch** [comment]
Also adjust `scripts/publish/npm_publish.sh` to use `dist/js/cjs/benchpress_bundle` folder.

File: gulpfile.js [hunk 6]
```diff
@@ -963,6 +968,29 @@ gulp.task('cleanup.builder', function(done) {
   });
 });
 
+gulp.task('benchpress.bundle', ['build.js.cjs'], function() {
+  var b = browserify({
```
Comment (tbosch) [inline gulpfile.js]: Move the implementation to `tools/build`, but leave the configuration (e.g. paths) here. See the usage of `bundler`

File: gulpfile.js [hunk 7]
```diff
@@ -963,6 +968,29 @@ gulp.task('cleanup.builder', function(done) {
...
+    entries: './dist/js/cjs/benchpress/index.js'
+  });
+  b.exclude('selenium-webdriver');
+  var packageJson = require('./dist/js/cjs/benchpress/package.json')
```
Comment (tbosch) [inline gulpfile.js]: Don't require it, as `require` caches. Use `fs.readFileSync` / JSON.parse

File: gulpfile.js [hunk 8]
```diff
@@ -963,6 +968,29 @@ gulp.task('cleanup.builder', function(done) {
...
+  for (var dep in packageJson.dependencies) {
+    //remove deps from package that we want to include in the bundle
+    if (dep == "reflect-metadata" || dep == "angular2" || dep == "traceur") {
```
Comment (tbosch) [inline gulpfile.js]: configure these in an object hash

Comment (tbosch) [inline gulpfile.js]: Use `===` instead of `==`

File: gulpfile.js [hunk 9]
```diff
@@ -963,6 +968,29 @@ gulp.task('cleanup.builder', function(done) {
...
+      delete packageJson.dependencies[dep];
+    } else {
+      b.exclude(dep);
+    }
+    mkdirp.sync('./dist/js/cjs/benchpress_bundle/');
+    require('fs').writeFileSync('./dist/js/cjs/benchpress_bundle/package.json', JSON.stringify(packageJson));
```
Comment (tbosch) [inline gulpfile.js]: Make require toplevel

File: gulpfile.js [hunk 10]
```diff
@@ -963,6 +968,29 @@ gulp.task('cleanup.builder', function(done) {
...
+  }
+  return b.bundle()
```
Comment (tbosch) [inline gulpfile.js]: Try to use `fs.writeFile`

File: modules/benchpress/index.ts [hunk 11]
```diff
@@ -1,2 +1,5 @@
-require(require('traceur').RUNTIME_PATH);
+require('reflect-metadata');
+require('traceur/bin/traceur-runtime');
 module.exports = require('./benchpress.js');
+// used for getting out of browserify's scope.
```
Comment (tbosch) [inline modules/benchpress/index.ts]: Add comment for bundle

**Comment by tbosch** [comment]
Also change npm-shrinkwrap file, see npm-shrinkwrap.md in main folder of angular2

File: gulpfile.js [hunk 12]
```diff
@@ -127,6 +127,20 @@ var CONFIG = {
...
+  packageJson: './dist/js/cjs/benchpress/package.json',
+  includes: [
+    'reflect-metadata',
+    'angular2',
+    'traceur'
+  ],
+  excludes: [
+    'selenium-webdriver'
+  ],
+  dest: './dist/js/cjs/benchpress_bundle/'
```
Comment (tbosch) [inline gulpfile.js]: Let's change this to `dist/build/benchpress_bundle`, similar to the other bundle tasks.

**Comment by tbosch** [comment]
Let's also use the bundle to run tests in CI, so that we are sure things work as expected:

1. change `build.js.cjs` in `gulpfile.js` to also run `benchpress.bundle`
1. change `protractor-shared.js` to * `require('./dist/build/benchpress_bundle')` instead of `require('./dist/js/cjs/benchpress/benchpress')`

**Comment by tbosch** [comment]
Reproduce locally:

1. `gulp build.js.cjs`
2. `gulp build.js.prod`
3. `protractor protractor-js.conf.js --specs=./dist/js/cjs/benchmarks/e2e_test/tree_perf.js

File: gulpfile.js [hunk 13]
```diff
@@ -124,10 +124,29 @@ var CONFIG = {
...
+  ],
+  excludes: [
+    'selenium-webdriver',
+    'rtts_assert',
+    'zone.js'
+  ],
+  ignore: [
+    'rx'
+  ],
+  dest: './dist/build/benchpress_bundle/'
```
Comment (tbosch) [inline gulpfile.js]: Use `CONFIG.benchpress_bundle` here as well.

File: protractor-shared.js [hunk 14]
```diff
@@ -1,6 +1,6 @@
 // load traceur runtime as our tests are written in es6
```
Comment (tbosch) [inline protractor-shared.js]: Maybe change to "Note: Requiring benchpress loads traceur-runtime and reflect-metadata as well"

File: tools/build/bundle.js [hunk 15]
```diff
@@ -23,3 +25,39 @@ module.exports.modify = function(srcs, concatName) {
...
+  b.bundle(function(err, buf) {
+    if (err) {
+      return cb(err);
+    }
+    var contents = buf.toString();
+    contents += 'module.exports = global.__benchpressExports;\n';
+    fs.writeFileSync(dest + '/index.js', contents);
+    cb(null);
+  });
+}
```
Comment (tbosch) [inline tools/build/bundle.js]: Add newline at end

**Comment by tbosch** [comment]
Also include licenses:
Loop over `incude` packages given in `gulpfile.js` and read out the `LICENSE` file from these packages and copy it over into the bundle folder with file pattern: `LICENSE.traceur`, ...
Fail if a package is listed in the `includes` but has no `LICENSE` file.

**Comment by tbosch** [comment]
You might want to use `require.resolve` (see https://nodejs.org/api/globals.html#globals_require_resolve) to find the path for a module.

**Comment by uber5001** [comment]
`require.resolve` gives the location of the module's "main" script. The license files are closer to `angular/node_modules/<packageName>/LICENSE`, with the exception of angular2.

**Comment by tbosch** [comment]
Landed in master as https://github.com/angular/angular/commit/c4e10ea9ac2f5f76d82c727a03965dd9fa3fbea5

**Pull request closed without merging**